fairlearn/reductions/exponentiated_gradient/exponentiated_gradient.py

```diff
@@ -0,0 +1,267 @@
+from collections import namedtuple
+
+print = functools.partial(print, flush=True)
+
+
+def _mean_pred(X, hs, weights):
+    """Return a weighted average of predictions produced by classifiers in hs"""
+    pred = pd.DataFrame()
+    for t in range(len(hs)):
+        pred[t] = hs[t](X)
+    return pred[weights.index].dot(weights)
+
+
+# TODO: decide whether we want a structured object instead of the namedtuple
+column_names = "best_classifier best_gap classifiers weights last_t best_t n_oracle_calls"
+expgrad_result = namedtuple("ExpgradResult", column_names)
+
+
+class ExponentiatedGradientResult:
+    def __init__(self, best_classifier, best_gap, classifiers, weights, last_t, best_t,
+                 n_oracle_calls):
+        self._best_classifier = best_classifier
+        self._best_gap = best_gap
+        self._classifiers = classifiers
+        self._weights = weights
+        self._last_t = last_t
+        self._best_t = best_t
+        self._n_oracle_calls = n_oracle_calls
+
+    def _predict_with_best_classifier(self, X):
+        return self._best_classifier(X)
+
+    def _asdict(self):
+        return {
+            "best_classifier": self._best_classifier,
+            "best_gap": self._best_gap,
+            "classifiers": self._classifiers,
+            "weights": self._weights,
+            "last_t": self._last_t,
+            "best_t": self._best_t,
+            "n_oracle_calls": self._n_oracle_calls
+        }
+
+
+class ExponentiatedGradient(ReductionsEstimator):
+    def __init__(self, estimator, disparity_metric, quality_metric, eps=0.01):
+        self._estimator = estimator
+        self._disparity_metric = disparity_metric
+        # TODO use quality metric
+        self._quality_metric = quality_metric
+        self._eps = eps
+        self._best_classifier = None
+        self._classifiers = None
+
+    def fit(self, X, y, aux_data=None, **kwargs):
+        # TODO: validate input data; unify between grid search an expgrad?
+        self._expgrad_result = exponentiated_gradient(X, aux_data, y, self._estimator,
+                                                      constraints=self._disparity_metric,
+                                                      eps=self._eps)
+        self._best_classifier = expgrad_result.best_classifier
+        self._classifiers = expgrad_result.classifiers
+        # TODO: figure out whether we should keep the remaining data of the result object
+
+    def predict(self, X):
+        return self._expgrad_result._predict_with_best_classifier(X)
+
+    def predict_proba(self, X):
+        raise NotImplementedError()
+
+    def posterior_predict(self, X):
+        raise NotImplementedError()
+
+    def posterior_predict_proba(self, X):
+        raise NotImplementedError()
+
+    def _expgrad_result(self):
+        # Just making this available since it's needed for some of the tests
+        return self._expgrad_result
+
+
+def exponentiated_gradient(X, A, y, estimator,
```

Comment:
Why not make this a member method?

Reply:
Good point. I've not had too much time to do that, and I want to first determine how much of this will be reusable for regression. If those are totally separate then there will be a lot of code. I'm not 100% convinced that all needs to land in that class. Maybe that's better, maybe not? I'm open to all options right now, but focused on a working solution.

Reply:
Fair enough.